cli/command/system/prune_progress.go

```diff
@@ -0,0 +1,173 @@
+package system
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/pkg/jsonmessage"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/streamformatter"
+)
+
+type pruneProgressChan struct {
+	isEventsListening chan bool
+	errorChan         chan error
+}
+
+func runPruneProgress(dockerCli command.Cli, progressChans *pruneProgressChan, options pruneOptions) {
+	eventChan := make(chan string, 1)
+	objCount, err := countObjectsToRemove(dockerCli, options)
+	if err != nil {
+		progressChans.errorChan <- err
+		return
+	}
+	go monitorRemovingEvents(dockerCli, eventChan, progressChans)
+	go outputPruningProgress(dockerCli, eventChan, objCount, progressChans)
+}
+
+func outputPruningProgress(dockerCli command.Cli, events <-chan string, numEvents int, progressStruct *pruneProgressChan) {
+	pipeReader, pipeWriter := io.Pipe()
+	progressOut := streamformatter.NewJSONProgressOutput(pipeWriter, false)
+
+	go func() {
+		var i int64 = 1
+		for s := range events {
+			prog := progress.Progress{
+				ID:         "Progress",
+				Action:     s,
+				Current:    i,
+				Total:      int64(numEvents),
+				HideCounts: true,
+			}
+			if err := progressOut.WriteProgress(prog); err != nil {
+				progressStruct.errorChan <- err
+				return
+			}
+			i++
+		}
+	}()
+
+	if err := jsonmessage.DisplayJSONMessagesToStream(pipeReader, dockerCli.Out(), nil); err != nil {
+		progressStruct.errorChan <- err
+		return
+	}
+}
+
+func countObjectsToRemove(dockerCli command.Cli, options pruneOptions) (int, error) {
```

Comment:
I need to look at this some more, but it seems unlikely that we can accurately count the objects that will be pruned. What happens if objects are created after this count runs, but before system prune starts pruning objects of that type?

Reply:
Unfortunately, we can't accurately predict the total number of objects to remove, in the case of an object created after the count runs.